watch/source.go

```diff
@@ -18,32 +18,42 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-package watch
+package xwatch
 
 import (
+	"errors"
 	"sync"
 
 	"github.com/m3db/m3x/close"
 	"github.com/m3db/m3x/log"
 )
 
-// SourcePollFn provides source data
-type SourcePollFn func() (interface{}, error)
+// ErrSourceClosed could be thrown from SourceInput to indicate that the Source should be closed
+var ErrSourceClosed = errors.New("source closed")
+
+// SourceInput provides data for Source
+type SourceInput interface {
+	// Poll will be called by Source for data, any backoff/jitter logic should be handled here
+	Poll() (interface{}, error)
+}
 
 // Source polls data by calling SourcePollFn and notifies its watches on updates
 type Source interface {
 	xclose.SimpleCloser
 
+	// Initialized returns a channel that blocks until the Source was initialized
+	Initialized() <-chan struct{}
 	// Watch returns the value and an Watch
 	Watch() (interface{}, Watch, error)
 }
 
 // NewSource returns a Source
-func NewSource(poll SourcePollFn, logger xlog.Logger) Source {
+func NewSource(input SourceInput, logger xlog.Logger) Source {
 	s := &source{
-		poll:   poll,
+		input:  input,
 		w:      NewWatchable(),
 		logger: logger,
+		ch:     make(chan struct{}),
```

Comment:
Perhaps `initCh` is better?  `ch` makes it sound like it actually might back the data or something.